sthgetter/sthgetter_test.go

```diff
@@ -0,0 +1,187 @@
+		SHA256RootHash:    mustB64Decode("VicMkhzrGNv+lNCwXRVHH0WniZuDg3IXhgPai5kyHdA="),
+		TreeHeadSignature: mustB64Decode("BAMARzBFAiEAs0GiYnPT5ZQJ2LGLhLmIXZXSLg+N+CxTkJL75tECEqgCIBZzJGyzH9h+IL63XCvRlfTKhLvzSxVicrT30+rwTSU0"),
+	}
+	// The public key for the Google Pilot Log.
+	b64PubKey = "MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEfahLEimAoz2t01p3uMziiLOl/fHTDM0YDOhBRuiBARsV4UvxG2LdNgoIGLrtCzWE0J5APC2em4JlvR8EEEFMoA=="
+	// A time less than 24 hours after the timestamp of validSTH.  If validSTH
+	// was received at this time, it would not be considered 'too old'.
+	validReceiveTime   = time.Date(2019, time.April, 10, 15, 0, 0, 0, time.UTC)
+	invalidReceiveTime = time.Date(2019, time.April, 11, 15, 0, 0, 0, time.UTC)
+)
+
+func TestCheckSTH(t *testing.T) {
+	// Create Log structure.
+	l, err := ctlog.New("https://ct.googleapis.com", "google_pilot", b64PubKey, 24*time.Hour)
+	if err != nil {
+		t.Fatalf("Unable to obtain Log metadata: %s", err)
+	}
+
+	// Create signature verifier.
+	sv, err := ct.NewSignatureVerifier(l.PublicKey)
+	if err != nil {
+		t.Fatalf("Couldn't create signature verifier: %s", err)
+	}
+
+	tests := []struct {
+		desc         string
+		sth          *ct.GetSTHResponse
+		receivedAt   time.Time
+		wantErrTypes []reflect.Type
+	}{
+		{
+			desc:       "valid",
+			sth:        validSTH,
+			receivedAt: validReceiveTime,
+		},
+		{
+			desc: "invalid signature",
+			// STH with TreeSize modified (set to 0) so that signature will not
+			// verify.
+			sth: &ct.GetSTHResponse{
+				Timestamp:         validSTH.Timestamp,
+				SHA256RootHash:    validSTH.SHA256RootHash,
+				TreeHeadSignature: validSTH.TreeHeadSignature,
+			},
+			receivedAt: validReceiveTime,
+			wantErrTypes: []reflect.Type{
+				reflect.TypeOf(&SignatureVerificationError{}),
+			},
+		},
+		{
+			desc:       "old STH",
+			sth:        validSTH,
+			receivedAt: invalidReceiveTime,
+			wantErrTypes: []reflect.Type{
+				reflect.TypeOf(&OldTimestampError{}),
+			},
+		},
+		{
+			desc: "invalid sig and old STH",
+			sth: &ct.GetSTHResponse{
+				Timestamp:         validSTH.Timestamp,
+				SHA256RootHash:    validSTH.SHA256RootHash,
+				TreeHeadSignature: validSTH.TreeHeadSignature,
+			},
+			receivedAt: invalidReceiveTime,
+			wantErrTypes: []reflect.Type{
+				reflect.TypeOf(&SignatureVerificationError{}),
+				reflect.TypeOf(&OldTimestampError{}),
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.desc, func(t *testing.T) {
+			sth, err := test.sth.ToSignedTreeHead()
+			if err != nil {
+				t.Fatalf("bad bad: %s", err)
+			}
+
+			errs := checkSTH(sth, test.receivedAt, sv, l)
+			if len(errs) != len(test.wantErrTypes) {
+				t.Fatalf("checkSTH(%v, %v, _, _) = %v (%d errors), want errors of types %v (%d errors)", sth, test.receivedAt, errs, len(errs), test.wantErrTypes, len(test.wantErrTypes))
+			}
+
+			// Slightly brittle test: Relies on the order of the returned slice
+			// of errors from checkSTH().
```

Comment:
If you wanted to remove the brittleness, you could do something like this:
```go
errTransformer := cmp.Transformer("types", func(err error) reflect.Type { return reflect.TypeOf(err) })
typeSorter := cmpopts.SortSlices(func(x, y reflect.Type) bool { return x.String() < y.String() })
if diff := cmp.Diff(test.wantErrTypes, errs, errTransformer, typeSorter); diff != "" {
  t.Errorf("checkSTH(%v, %v, _, _); errors diff -want +got\n%s", sth, test.receivedAt, diff)
}
```
That'd also remove the need to check the length of `errs` above. It is a less readable though, since it uses a more advanced feature of `cmp`.

Reply:
So my initial reaction was "Yeah why not, that's pretty fun", but then I spent some time reading the documentation, go confused about how the magic of Diff Options work, and so am going to leave it as is for now, but may change it later if I get this stuff clear in my head.... that cool?

Reply:
Yea that's fine. It's probably not worth the added complexity unless you find that this test breaks in the future. I mostly provided that code because it can be handy to know about some of those options (`cmpopts.SortSlices()` is particularly helpful when you don't want to assume ordering).